### ISSUE TITLE
No precompiled python package for alpine Linux 

### ISSUE THREAD

**Issue by sam-qordoba** [grpc/grpc#18150]
### What version of gRPC and what language are you using?
 
1.18.0
 
### What operating system (Linux, Windows,...) and version?
 
Arch
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 
Trying to use pip...
 
### What did you do?
If possible, provide a recipe for reproducing the error. Try being specific and include code snippets if helpful.

minimal example: see [this Dockerfile](https://github.com/sam-qordoba/grpci_slow/blob/master/Dockerfile)
 
### What did you expect to see?
 
Expected grpcio to be pulled from pip as a package, and take a reasonable amount of time.
 
### What did you see instead?

The container takes about 5 minutes to build, over 90% of which is spent building grpcio

```sh
Building wheels for collected packages: grpcio
  Building wheel for grpcio (setup.py): started
  Building wheel for grpcio (setup.py): still running...
  Building wheel for grpcio (setup.py): still running...
  Building wheel for grpcio (setup.py): still running...
  Building wheel for grpcio (setup.py): finished with status 'done'

...

real	4m44.918s
user	0m0.151s
sys	0m0.158s
```

Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

See [TROUBLESHOOTING.md](https://github.com/grpc/grpc/blob/master/TROUBLESHOOTING.md) for how to diagnose problems better.
 
### Anything else we should know about your project / environment?

I know there have been closed issues like this... You can see the Dockerfile here https://github.com/sam-qordoba/grpci_slow/blob/master/Dockerfile ... it's as minimal can be.

**Comment by nicolasnoble** [comment]
There is a conscious decision on our side to have a list of things that we are supporting. We can't produce precompiled binaries for every single combination of platforms out there. The compilation from source fallback is the best way to solve this, but this means that it can take a while.

**Comment by sam-qordoba** [comment]
Sorry, I did not realize it was intentional. Let's say I wanted to produce a precompiled binary for the platform(s) of my choice... Do I have to read and understand all of `setup.py` in root of this repo, or is there a fairly straightforward way to maintain an up-to-date binary for Alpine Linux? Thanks for the help.

**Comment by mehrdada** [comment]
Unless it's a recent change, I don't believe PyPI repo even allows for uploading packages that aren't `manylinux1` and distro-specific.

**Comment by mehrdada** [comment]
@sam-qordoba `pip install grpcio` should automatically work on alpine if you have the C toolchain installed (there are tests to ensure it works in [an Alpine docker image](https://github.com/grpc/grpc/blob/master/tools/dockerfile/test/python_alpine_x64/Dockerfile)). Doesn't it? Shouldn't require any understanding of `setup.py`.  Building a precompiled package should also be [straightforward](https://github.com/grpc/grpc/blob/master/tools/run_tests/artifacts/build_artifact_python.sh#L40) with general familiarity with distribtools, and you can host it in your own pip repository and explicitly point your clients to it.

**Comment by sam-qordoba** [comment]
@mehrdada Correct, it does work. It just always builds from source. I thought that was a bug, but @nicolasnoble let me know that it is a conscious decision. Thank you for the additional feedback.

**Comment by nicolasnoble** [comment]
As @mehrdada points out, this may also be a limitation in what we can do with PyPI.

**Comment by danielfaust** [comment]
I just want to point out that explicitly flagging the installation as manylinux1 compatible, like so:
```  
RUN apk --update --no-cache add \
[...]
python3 python3-dev py3-pip && \
echo 'manylinux1_compatible = True' > /usr/lib/python3.6/site-packages/_manylinux.py && \
pip3 install --upgrade pip && \
pip3 install firebase-admin \
[...]
```
This can end up with a working image and speed up the build time dramatically. In my case it was a little bit over 10 minutes before to down to slightly less than 1 minute after the change and all is still working.

It's really worth it giving it a try.

**Comment by gnossen** [comment]
It's worth noting that, due to its usage of musl libc, [Alpine is not manylinux compatible](https://github.com/pypa/manylinux/issues/37). It might "mostly work," but an incompatibility could bite you down the road.

**Comment by lidizheng** [comment]
Compiling Python packages actually adds a lot of size to the final Docker image. IMO if the package installation process end up compiling many things on the fly, it might be a signal to use another base image (like [busybox](https://hub.docker.com/_/busybox)).

I don't think we can do much here, since alpine Linux is not manylinux compatible, and we can't upload binary wheel for alpine to PyPI. Do you think we should close this issue? @gnossen

**Comment by gnossen** [comment]
I think it's useful as a tracking issue. It's clear that there's demand for prebuilt Alpine images (presumably due to its usage in the Docker community). We're blocked on https://github.com/pypa/packaging-problems/issues/69 (Note that @theacodes has already mentioned us on that thread.) Once Pypa puts a framework in place that enables us to release precompiled packages for Alpine, I absolutely think we should.

**Comment by ye** [comment]
> I just want to point out that explicitly flagging the installation as manylinux1 compatible, like so:
> 
> ```
> RUN apk --update --no-cache add \
> [...]
> python3 python3-dev py3-pip && \
> echo 'manylinux1_compatible = True' > /usr/lib/python3.6/site-packages/_manylinux.py && \
> pip3 install --upgrade pip && \
> pip3 install firebase-admin \
> [...]
> ```
> 
> can end up with a working image and speed up the build time dramatically. In my case it was a little bit over 10 minutes before to down to slightly less than 1 minute after the change and all is still working.
> 
> It's really worth it giving it a try.
> 
> A couple of months later (August 2019): Alpine 3.10 switched to Python 3.7 and it was no longer possible to use this workaround. I will stop using this approach.

Taking the hint there, I made it work on python 3.7 after poking around the new file structure in the newer image.

`_manylinux.py` location changed. So change it as follow will make it work :) Again super hack-ish but it gets the build done.

```bash
echo 'manylinux1_compatible = True' > /usr/local/lib/python3.7/site-packages/_manylinux.py && \
```

**Comment by ChazZeromus** [comment]
Was browsing the setup files for grpcio and found this https://github.com/grpc/grpc/blob/master/src/python/grpcio/_parallel_compile_patch.py#L25

In addition to base images you can override the build jobs by adding something like `ENV GRPC_PYTHON_BUILD_EXT_COMPILER_JOBS 16` to your docker file right before pip. Builds much faster this time!

**Comment by whalesalad** [comment]
@ChazZeromus that change combined with increasing the heft of my buildbox (using AWS Codebuild which uses Docker under the hood) has dramatically improved my performance. Thank you immensely for the tip!

**Comment by ye** [comment]
@danielfaust No clue. I am still sticking around with 3.7 for now as most the dependent pip packages are still working against 3.7 (would probably work on 3.8 but I didn't get time to check.)

**Comment by jtattermusch** [comment]
We acknowledge this to be a valid feature request but since no one has come forward to implement this, we are going to close it for now while maintaining the label/Help wanted to indicate that this can be picked up as future work.
